File: cloudwatch_log_trigger/cloudwatch_log_sender.js
var zlib = require('zlib');
var util = require('util');
var AWS = require('aws-sdk');


const PATTERNS = [
    'REPORT', 'Unable to import module', 'Task timed out',
    'Process exited before completing', 'Traceback',
    'module initialization error:', 'errorMessage'
];
const MAX_STR_SIZE = 100;

const KINESIS_CLIENT = new AWS.Kinesis({
    region: process.env.EPSAGON_REGION,
    accessKeyId: process.env.EPSAGON_AWS_ACCESS_KEY_ID,
    secretAccessKey: process.env.EPSAGON_AWS_SECRET_ACCESS_KEY
});

function epsagon_debug(error) {
    if ((process.env.EPSAGON_DEBUG || '').toUpperCase() === 'TRUE') {
        // eslint-disable-next-line no-console
        console.log(error);
    }
}

module.exports.forwardLogsData = function forwardLogsData(logsData) {
    return new Promise((resolve, reject) => {
        epsagon_debug('Attempting to forward logs');

        var zippedInput = new Buffer.from(logsData, 'base64');
        epsagon_debug(util.format('Size before compression %d bytes', zippedInput.length));
        zlib.gunzip(zippedInput, function (e, buffer) {
            if (e) {
                epsagon_debug(e);
                resolve();
                return;
            }

            epsagon_debug(util.format('Size after decompression %d bytes', buffer.length));

            var awslogsData = JSON.parse(buffer.toString('utf-8'));

            if (awslogsData.messageType !== 'DATA_MESSAGE') {
                epsagon_debug('Message other than DATA_MESSAGE received');
                resolve();
                return;
            }

            var forwadedMsgs = [];

            epsagon_debug(util.format('Scanning %d lines', awslogsData.logEvents.length));

            awslogsData.logEvents.forEach(function (log, idx, arr) {
                PATTERNS.forEach(pattern => {
                    if (log.message.slice(0, MAX_STR_SIZE).indexOf(pattern) >= 0) {
                        forwadedMsgs.push(log)
                    }
                })
            });

            epsagon_debug('Done scanning');

            if (forwadedMsgs.length !== 0) {
                awslogsData.logEvents = forwadedMsgs;
                awslogsData.subscriptionFilters = [];
                awslogsData.subscriptionFilters.push(util.format('Epsagon#%s#%s', awslogsData.owner, process.env.AWS_REGION));

                zlib.gzip(new Buffer.from(JSON.stringify(awslogsData), 'ascii'), function (e, buffer) {
                    if (e) {
                        epsagon_debug('Failed compressing result');
                        resolve();
                        return;
                    }

                    // Forward to Epsagon Kinesis
                    try {
                        var params = {
                          Data: buffer,
                          PartitionKey: awslogsData.logStream,
                          StreamName: process.env.EPSAGON_KINESIS_NAME
                        };

                        epsagon_debug(util.format('About to forward %d records', forwadedMsgs.length));

                        KINESIS_CLIENT.putRecord(params, function(err, data) {
                          epsagon_debug('Record sent');
                          epsagon_debug(data);
                          if (err) {
                            epsagon_debug(err);
                          }
                          resolve();
                        });
                    }
                    catch (e) {
                        epsagon_debug('Got error in forwarding');
                        epsagon_debug(e);
                        resolve();
                    }
                })
            }
        });
    });
}

// Lambda handler for forwarding logs either from CloudWatch logs trigger, or from a
//  Kinesis stream of a CloudWatch subscription
module.exports.forwardLogsLambdaHandler = function forwardLogsLambdaHandler(event, context) {
    if (event.awslogs) {
        return module.exports.forwardLogsData(event.awslogs.data);
    }

    if (event.Records) {
        event.Records.forEach(function(record) {
            module.exports.forwardLogsData(record.kinesis.data);
        });
    }
};

File: cloudwatch_log_trigger/cloudwatch_log_sender.py
"""Epsagon Logs Parser, parse CloudWatch Logs event"""

import json
import gzip
import base64
import os
import re
import traceback
import boto3

FILTER_PATTERNS = (
    'REPORT', 'Task timed out', 'Process exited before completing', 'Traceback',
    'module initialization error:', 'Unable to import module', 'errorMessage',
    '.java:0', '.java:1', '.java:2', '.java:3', '.java:4', '.java:5', '.java:6',
    '.java:7', '.java:8', '.java:9'
)


AWS_SECRET = os.environ.get('EPSAGON_AWS_SECRET_ACCESS_KEY').strip()
AWS_KEY = os.environ.get('EPSAGON_AWS_ACCESS_KEY_ID').strip()
REGION = os.environ.get('EPSAGON_REGION').strip()
KINESIS_NAME = os.environ.get('EPSAGON_KINESIS_NAME').strip()
REGEX = re.compile(
    '|'.join([f'.*{pattern}.*' for pattern in FILTER_PATTERNS]),
    re.DOTALL
)

kinesis = boto3.client(
    'kinesis',
    aws_access_key_id=AWS_ID,
    aws_secret_access_key=AWS_KEY,
    region_name=REGION,
)


def epsagon_debug(message):
    if os.getenv('EPSAGON_DEBUG', '').upper() == 'TRUE':
        print(message)


def filter_events(record_data):
    """
    Filter events relevant for Epsagon.
    :param record_data: Record data that holds the vents.
    :return: dict / None.
    """
    record = None
    if record_data['messageType'] == 'DATA_MESSAGE':
        original_events = record_data['logEvents']
        partition_key = record_data['logStream']
        record_data['subscriptionFilters'] = (
            [f'Epsagon#{record_data["owner"]}#{os.getenv("AWS_REGION")}']
        )
        events = []
        epsagon_debug(f'Found total of {len(original_events)} events')
        epsagon_debug(f'Original events: {original_events}')
        for event in original_events:
            if REGEX.match(event['message']) is not None:
                events.append(event)
        epsagon_debug(f'Filtered total of {len(events)} events.')
        if events:
            record_data['logEvents'] = events
            record = {
                'Data': gzip.compress(json.dumps(record_data).encode('ascii')),
                'PartitionKey': partition_key
            }
    return record


def forward_logs_to_epsagon(event):
    """
    Send filtered CloudWatch logs to Epsagon Kinesis.
    :param event: The triggered event from CloudWatch logs.
    """
    try:
        record_data = json.loads(
            gzip.decompress(
                base64.b64decode(event['awslogs']['data'])
            )
        )
        filtered_event = filter_events(record_data)
        if not filtered_event:
            epsagon_debug('No logs match')
            return False

        original_access_key = os.environ.pop('AWS_ACCESS_KEY_ID')
        original_secret_key = os.environ.pop('AWS_SECRET_ACCESS_KEY')
        original_region = os.environ.pop('AWS_REGION')
        os.environ['AWS_ACCESS_KEY_ID'] = AWS_KEY
        os.environ['AWS_SECRET_ACCESS_KEY'] = AWS_SECRET
        os.environ['AWS_REGION'] = REGION
        try:
            kinesis.put_record(
                StreamName=KINESIS_NAME,
                Data=filtered_event['Data'],
                PartitionKey=filtered_event['PartitionKey'],
            )
        finally:
            os.environ['AWS_ACCESS_KEY_ID'] = original_access_key
            os.environ['AWS_SECRET_ACCESS_KEY'] = original_secret_key
            os.environ['AWS_REGION'] = original_region

    except Exception as err:
        epsagon_debug('Encountered error: {}'.format(err))
        epsagon_debug(traceback.format_exc())

    return True
